File: F1Levier/app/src/main/java/com/djira/f1levier/dao/ParticipantDao.java
package com.djira.f1levier.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;

import com.djira.f1levier.entity.Equipe;
import com.djira.f1levier.entity.Participant;

import java.util.List;

import static androidx.room.OnConflictStrategy.REPLACE;

@Dao
public interface ParticipantDao {
    //Insert query
    @Insert(onConflict = REPLACE)
    void insert(Participant participant);

    //Delete query
    @Delete
    void delete(Participant participant);

    //Delete all query
    @Delete
    void reset(List<Participant> participant);

    //Update query
    @Query("UPDATE Participant SET name = :sName, level = :sLevel WHERE ID = :sID")
    void update(int sID, String sName, Integer sLevel);

    //Update query
    @Query("UPDATE Participant SET equipe_id = :sEquipeID WHERE ID = :sID")
    void updateEquipe(int sID, Integer sEquipeID);

    //Update query
    @Query("UPDATE Participant SET equipe_id = 0 WHERE equipe_id != 0")
    void deleteFromEquipe();

    //Update query
    @Query("UPDATE Participant SET t1 = :sT1, t1 = :sT2, ps = :sPs, t3 = :sT3, t4 = :sT4 WHERE ID = :sID")
    void updateTemps(int sID, int sT1,int sT2,int sPs,int sT3,int sT4 );

    //Get all data query
    @Query("SELECT * FROM Participant")
    List<Participant> getAll();

    //Get all data query
    @Query("SELECT * FROM Participant ORDER BY level ASC")
    List<Participant> getAllByLevel();

    //Get participant by equipe
    @Query("SELECT Participant.* FROM Participant " +
            "INNER JOIN Equipe " +
            "ON Participant.equipe_id = Equipe.ID WHERE Equipe.name = :eqName")
    List<Participant> getAllByEquipe(String eqName);

    //Get participant by equipe
    @Query("SELECT Participant.*, " +
            "SUM(Participant.t1 + Participant.t2 + Participant.ps + Participant.t3 + Participant.t4) AS temps FROM Participant " +
            "INNER JOIN Equipe " +
            "ON Participant.equipe_id = Equipe.ID WHERE Equipe.ID = :eqID " +
            "GROUP BY temps, Participant.ID, Participant.name, " +
            "Participant.level, Participant.t1, Participant.t2," +
            " Participant.ps, Participant.t3, Participant.t4, Participant.equipe_id " +
            "ORDER BY temps ASC" +
            "")
    List<Participant> getAllByEquipeID(Integer eqID);

    //Get participant by equipe
    @Query("SELECT Equipe.* FROM Participant " +
            "INNER JOIN Equipe " +
            "ON Participant.equipe_id = Equipe.ID WHERE Participant.ID = :sID")
    Equipe getEquipeByParticipantID(Integer sID);

}

File: Projet/app/app/src/main/java/com/codep25/codep25/model/storage/dao/ParticipantDao.kt
package com.codep25.codep25.model.storage.dao

import androidx.room.*
import com.codep25.codep25.model.entity.Participant
import com.codep25.codep25.model.entity.ParticipantWithTeam

@Dao
interface ParticipantDao {

    @Query("SELECT * FROM participants")
    fun getParticipants(): List<Participant>

    @Query("SELECT COUNT(*) FROM participants")
    fun getNumberOfParticipants(): Int

    @Query("SELECT * FROM participants WHERE id_participant == :id")
    fun getParticipantById(id: Long): Participant

    @Query("SELECT * FROM participants WHERE id_participant IN (:ids)")
    fun getParticipantByIds(vararg ids: Long): List<Participant>
    
    @Query("SELECT * FROM participants WHERE participant_level > :lvl")
    fun getParticipantsOverLevel(lvl: Int): List<Participant>

    @Query("SELECT * FROM participants ORDER BY participant_level ASC")
    fun getParticipantsOrderedBySkill(): List<Participant>

    @Insert
    fun insert(participant: Participant)

    @Insert
    fun insertAll(vararg participant: Participant): List<Long>

    @Update
    fun update(participant: Participant)

    @Update
    fun updateAll(vararg participant: Participant)
    
    @Query("UPDATE participants SET participant_level = :lvl WHERE participant_level > :lvl")
    fun updateMaxLevel(lvl: Int)

    @Query("DELETE FROM participants where id_participant IN (:ids)")
    fun delete(vararg ids: Long)

    @Query("DELETE FROM participants")
    fun clearParticipants()
}